packages/gatsby/src/utils/page-data.ts

```diff
@@ -69,10 +79,83 @@ export async function writePageData(
   store.dispatch({
     type: `ADD_PAGE_DATA_STATS`,
     payload: {
-      filePath,
+      filePath: outputFilePath,
       size: pageDataSize,
     },
   })
 
-  await fs.outputFile(filePath, bodyStr)
+  await fs.outputFile(outputFilePath, bodyStr)
+  return body
+}
+
+let isFlushPending = false
+let isFlushing = false
+
+export function isFlushEnqueued(): boolean {
+  return isFlushPending
+}
+
+export async function flush(): Promise<void> {
+  if (isFlushing) {
+    // We're already in the middle of a flush
+    return
+  }
+  isFlushPending = false
+  isFlushing = true
+  const { pendingPageDataWrites, components, pages, program } = store.getState()
+
+  const { pagePaths, templatePaths } = pendingPageDataWrites
+
+  const pagesToWrite = Array.from(templatePaths).reduce(
+    (set, componentPath) => {
+      const templateComponent = components.get(componentPath)
+      if (templateComponent) {
+        templateComponent.pages.forEach(set.add.bind(set))
+      }
+      return set
+    },
+    new Set(pagePaths.values())
+  )
+
+  for (const pagePath of pagesToWrite) {
+    const page = pages.get(pagePath)
+
+    // It's a gloomy day in Bombay, let me tell you a short story...
+    // Once upon a time, writing page-data.json files were atomic
+    // After this change (#24808), they are not and this means that
+    // between adding a pending write for a page and actually flushing
+    // them, a page might not exist anymore щ（ﾟДﾟщ）
+    // This is why we need this check
+    if (page) {
+      const body = await writePageData(
+        path.join(program.directory, `public`),
+        page
+      )
+
+      if (program?._?.[0] === `develop`) {
+        websocketManager.emitPageData({
+          ...body.result,
+          id: pagePath,
+          result: {
+            data: body.result.data,
+            pageContext: body.result.pageContext,
+          },
+        })
+      }
+    }
+  }
+
+  store.dispatch({
+    type: `CLEAR_PENDING_PAGE_DATA_WRITES`,
+  })
```

Comment:
@sidharthachatterjee we introduced a bug with this (thank @TylerBarnes for raising this issue)

we generate `pagesToWrite` at the beginning of function (sync), and then the loop is async, which means that some concurrent code can add pending pages/templates while the loop is running and after that we clear all pending data writes which means that pending writes added mid function run are never handled.

What we did with Tyler to debug here is we rerun code used to generate `pagesToWrite` after the loop and realized that that it doesn't match with one we generated at the beginning of the function.

I think we either need to make this entire flush sync (yikes) or add some listener (e.g. `emitter.on("ADD_PENDING_PAGE_DATA_WRITE", ...)` that would flip some switch and rerun `flush` or something before dispatching `CLEAR_PENDING_PAGE_DATA_WRITES`

Reply:
For context this bug makes data updates break for WP due to missing page data but only on some sites. It's not reproducible on the default starter and seemingly needs a more complex site to reproduce

Reply:
Not sure if that would work - but maybe we should dispatch "clear pending" action immediately after constructing `pagesToWrite` (as code is sync up to this point), so after the loop we could just check if there is anything pending (meaning it was added since the clear) and then re-run pending? This way we won't need to listen to `ADD_PENDING_*` actions or make things sync?

Reply:
maybe a good use-case for a generator? after each iteration check to see if anything has been added and put it on the end of the processing queue